i18n tooling, JavaScript i18n and added the Japanese language

OmarIthawi opened the pull request:
We'd like to add Japanese as a supported language, also supported JavaScript translations via 
[`django-statici18n`](https://github.com/zyegfryed/django-statici18n/pull/41).

Before you continue, take a look at [edx/i18n-tools](https://github.com/edx/i18n-tools) and [edx/cookiecutter-django-app](https://github.com/edx/cookiecutter-django-app).

### Summary of the Changes
 - Added a Makefile to help manage the translations. The file is copied and modified from: [edx/cookiecutter-django-app](https://github.com/edx/cookiecutter-django-app)
 - Created a `locale` link to `poll/translations` to allow the `i18n_tools` find the po/mo files
 - Added `poll/translations/config.yaml` which works as the configuration file of the `i18n_tools` command 
 - Added `requirements-dev.txt` to pin development packages. Used for now only for translations.

### TODO

 - [x] Make Travis happy
 - [x] Support handlebars template as `djangojs.po`
 - [x] Document how to update the strings properly from Transifex
 - [x] Manually test that the ~Japanese~ the dummy Esperanto translation is working
 - [ ] Check the other open pull request for changes to cherry-pick: https://github.com/open-craft/xblock-poll/pulls

antoviaque commented:
@OmarIthawi That's great! Thank you for contributing this.

@bradenmacdonald The review on this could likely be part of the i18n project that's about to start?

OmarIthawi commented:
@antoviaque @bradenmacdonald Thank you :) 

Looking forward for your review.

OmarIthawi commented:
Thanks @ahmedaljazzar! I'll address your comments by the end of (my) today.

OmarIthawi commented:
@ahmedaljazzar I've updated both of the docs of this PR and the README.md. Please let me know if that will do.

Just an FYI, we're still working on the actual Japanese translations on Transifex.

ahmedaljazzar commented on Makefile:
```diff
@@ -0,0 +1,59 @@
+
+## Localization targets
+
+extract_translations: ## extract strings to be translated, outputting .po files
```
Why do we need to execute this command inside the poll directory? 
When I run it in a clean-configured repo, I get the following error:
```
CommandError: Unable to find a locale path to store translations for file __init__.py
```
However, running it without changing the directory just works perfect.

OmarIthawi replied:
Done

ahmedaljazzar commented on Makefile:
```diff
@@ -0,0 +1,59 @@
+push_translations: extract_translations ## push source translation files (.po) from Transifex
+	tx push -s
+
+dummy_translations: ## generate dummy translation (.po) files
```
And the same here.

ahmedaljazzar commented on Makefile:
```diff
@@ -0,0 +1,59 @@
+compile_translations: ## compile translation files, outputting .mo files for each supported language
+	django-admin compilemessages
+
+detect_changed_source_translations:
```
The same comment as above goes here, but with the following error:
```
cd poll && i18n_tool changed
Traceback (most recent call last):
  File "/.virtualenvs/xblock-utils/bin/i18n_tool", line 11, in <module>
    sys.exit(main())
  File "/.virtualenvs/xblock-utils/lib/python2.7/site-packages/i18n/main.py", line 60, in main
    return module.main()
  File "/.virtualenvs/xblock-utils/lib/python2.7/site-packages/i18n/__init__.py", line 50, in __call__
    self.configuration = config.Configuration(filename=args.config, root_dir=root_dir)
  File "/.virtualenvs/xblock-utils/lib/python2.7/site-packages/i18n/config.py", line 37, in __init__
    self._config = self.read_config(self._filename)
  File "/.virtualenvs/xblock-utils/lib/python2.7/site-packages/i18n/config.py", line 62, in read_config
    raise Exception("Configuration file cannot be found: %s" % filename)
Exception: Configuration file cannot be found: /Projects/OpenCraft/xblock/xblock-poll/poll/conf/locale/config.yaml
make: *** [detect_changed_source_translations] Error 1
```

OmarIthawi replied:
Done

OmarIthawi commented:
Hey @ahmedaljazzar! Happy Eid!

Just saw your review. Will get back to you soon.

OmarIthawi commented on README.md:
```diff
@@ -242,14 +242,25 @@ names in the django settings using the `XBLOCK_POLL_EXTRA_VIEW_GROUPS` setting,
 For information about working with translations, see the [Internationalization Support](http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_lms.html#internationalization-support) section of the [Open edX XBlock Tutorial](https://xblock-tutorial.readthedocs.io/en/latest/).
 
-## Downloading translations from Transifex
+## Working wht Translations and Transifex
```
Done

OmarIthawi commented on Makefile:
```diff
@@ -0,0 +1,59 @@
+	tx pull -af --mode reviewed
+	make compile_translations
+
+push_translations: extract_translations ## push source translation files (.po) from Transifex
```
Done

OmarIthawi commented:
@ahmedaljazzar I've addressed your comments. Please take a look and let me know.

The Japanese translation is still pending though.

OmarIthawi commented:
@ahmedaljazzar Here's a quick overview of the latest changes:

After your review I've done both, which mainly address your comments:
 - Correct typos
 - Address review comments

Later on I've also made:
 - Fixup source file name from `text.po` to `django.po`
 - Generated dummy translations so I can test the package locally
 - Add translations to Travis checks, so we ensure that the errors won't come up again out of nowhere

OmarIthawi commented:
Ah! Just realized that most of this XBlock is written with Handlebars, and Django thinks that they're a Python templates!

A good amount of refactoring is needed for this. Hold on a bit.

ahmedaljazzar commented:
Cool, @OmarIthawi, just ping me when it's ready and the Travis CI is happy again with the changes.

OmarIthawi commented:
Thanks @ahmedaljazzar!

I need a some sanity checks about my approach in translating JavaScript. I think that there's no XBlock (not even ORA2) supports JavaScript translations. While a lot of those XBlocks includes `djangojs.po` it's being ignored.

I've fixed this issue by converting the `djangojs.po` file into `djangojs.json` and loaded it as an XBlock JavaScript resource (`var PollCatalog`) which I then consume in the `PollCommonUtil` imitating how Django consumes its catalog.

I'd like to know whether I went too far with my approach. That is if there's a proper XBlock i18n for JavaScript files, only for Python files and Django templates.

I think @pomegranited have introduced such support for the Drag and Drop XBlock, but looks like this XBlock doesn't support JavaScript i18n either.

bradenmacdonald commented:
@pomegranited  would appreciate your comments on his approach ^

@OmarIthawi we have been doing work to support XBlock JS i18n. See:
* https://github.com/zyegfryed/django-statici18n/pull/41 (required, but not yet included in edx-platform)
* https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/156 and https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/172 which use that approach

pomegranited commented:
Thank you for your work here @OmarIthawi ! Some background to expand on the above comments:

> I've fixed this issue by converting the djangojs.po file into djangojs.json and loaded it as an XBlock JavaScript resource (var PollCatalog) which I then consume in the PollCommonUtil imitating how Django consumes its catalog.

That is basically what the [`django-statici18n`](https://github.com/zyegfryed/django-statici18n) library does -- it compiles `.po` files into `.js` files, which expose the translation library via Javascript `gettext` and `ngettext` methods.  EdX platform currently uses `django-statici18n` for its own JS translations, but since the generated JS by default provides global `gettext` methods, we submitted https://github.com/zyegfryed/django-statici18n/pull/41 to namespace the individual XBlock translations so they stay isolated from the platform and other XBlocks.

(We haven't bumped the platform version of `django-statici18n` because strictly speaking, the edxapp virtualenv doesn't need this new version.  The XBlock `.js` translation files are to be generated independently of the platform and need to be pushed into the XBlock deployment, but once that is done, those generated `.js` files work fine on their own.)

So I'd recommend doing the same thing here -- using [`django-statici18n`](https://github.com/zyegfryed/django-statici18n) to generate your JS files for you, instead of manually maintaining JSON translation resources.

OmarIthawi commented:
@pomegranited @bradenmacdonald That's much better! I'll switch to `django-statici18n`.

OmarIthawi commented:
@ahmedaljazzar I've refactored this PR to use `django-statici18n`. It's much cleaner now.

However I need to make Travis happy again, and clean up artifacts from previous attempts.

ahmedaljazzar commented:
Perfect! Thanks @OmarIthawi. Will take a look into it very soon.

OmarIthawi commented:
@ahmedaljazzar The PR is ready for another review, probably the last one.

The main bulk of the PR is kept into a single commit, with a couple of minor fixups after it. I'll squash after your review.

OmarIthawi commented on Makefile:
```diff
@@ -0,0 +1,120 @@
+	rm -fr *.egg-info
+
+quality: ## check coding style with pycodestyle and pylint
+	pep8 poll --max-line-length=120
```
Updated, added pycodestyles (pep8 successor) and updated the reqs.

OmarIthawi commented on Makefile:
```diff
@@ -0,0 +1,120 @@
+	echo > locale/en/LC_MESSAGES/textjs.po  # Ensure it's empty
+	# The sort to avoid bash arbitrary file order
+	ls poll/public/handlebars/*.handlebars | sort \
+	    | xargs npx xgettext-template --from-code utf8 \
```
I think it's a [standard part of nodejs](https://medium.com/@maybekatz/introducing-npx-an-npm-package-runner-55f7d4bd282b), is it available in your device?

OmarIthawi commented:
Thanks for your comments @ahmedaljazzar, I've addressed all of your comments. Except for `npx`, which should be installed by default.

OmarIthawi commented:
Thanks for the detailed review @ahmedaljazzar!

> I tested your work and there's one tiny thing that's not working for MAC users that I don't think it's a blocker which is the make target install_dev_firefox. It looks like making this command a cross-platform one is not straight forward as I thought and requires a little bit of a time since Firefox only provides a .dmg files for Darwin and the installation process is going to be different in this case. 
> However, we can include a note in the README.md file states that to run tests locally please note that you need A major.minor version of Firefox installed on your machine and here's the link. That's more convenient and easier to manage than building a script to do the process. You have the choice to keep the make target for Linux users or to remove it.

I've done the following:
 - Renamed the make commands to clarify that it's for Linux.
 - Updated the README to reflect those changes

> For npx that's true, I ran the commands on npm version 3 and it's only included starting from npm version 5. Maybe we can add this as note in the README.md as well?
I've removed npx in favor of the longer form. The command `$ make extract_translations` should work without requiring additional installs.

> Ran the ignored command that extracts JS translations `$ django-admin makemessages -l en -v1 -d djangojs -e js` and found out that is not working properly, but when I get red of \x04 from textjs.js files it did. I’m not going to consider this as a blocker for now because it’s not under the scope of this PR and we're not extracting JS strings now.

Thanks for doing such detailed test. Just in case we needed it in the future, this is the full command that will work:
```
$ django-admin makemessages -l en -v1 -d djangojs -e js --ignore node_modules --ignore translations
```

`django-admin` gets confused about the `textjs.js` files, it shouldn't be processing them in the first place.

Please take another look and let me know.

ahmedaljazzar commented:
Thanks @OmarIthawi! 👍

I tested this: Verified that the change works in both RTL and LTR mode, in different languages, and in a recent edx-platform.
